# ![MethBrain_small.png](src/cnsmeth/images/MethBrain_small.png)CNSmeth

[![PyPI - Version](https://img.shields.io/pypi/v/methnicegui.svg)](https://pypi.org/project/methnicegui)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/methnicegui.svg)](https://pypi.org/project/methnicegui)

-----

**Table of Contents**

- [Installation](#installation)
- [Usage](#usage)
- [About](#about)
- [License](#license)


## Installation

We recommend installing 'CNSmeth' using the following conda yml file included in the repository: [cnsmeth.yml](cnsmeth.yml)

This will install all the required dependencies including R and Python packages as well as readfish and ont-pyguppy-client-lib.

However, in this version of cnsmeth, you will need to install the tool from source. See below for installation details.

The contents of this file are:

```console
name: cnsmeth
channels:
  - bioconda
  - conda-forge
  - defaults
dependencies:
  - git-lfs
  - r-base
  - bioconda::bioconductor-genomicranges
  - r-optparse
  - r-data.table
  - conda-forge::r-ranger
  - r-matrixStats
  - r-glmnet
  - python=3.8
  - pip
  - pip:
    - readfish
    - ont-pyguppy-client-lib
    #- git+https://github.com/LooseLab/cnsmeth

```
then to create the environment:

```console
conda env create -f cnsmeth.yml
```

To activate the environment:

```console
conda activate cnsmeth
```

To install the tool from source:

```console
git clone https://github.com/LooseLab/cnsmeth/
cd cnsmeth
git lfs install
git lfs pull
git submodule update --init --recursive
pip install .
```


## Usage

```console
‚ùØ cnsmeth --help
Usage: cnsmeth [OPTIONS] [WATCHFOLDER] [OUTPUT]

  Entrypoint for when GUI is launched directly. :return: None

Options:
  --port INTEGER                  Port for GUI
  --threads INTEGER               Number of threads available.
  --simtime BOOLEAN               If set, will simulate the addition of
                                  existing files to the pipeline based on read
                                  data.
  --showerrors BOOLEAN            If set, will display all errors in running
                                  R.
  --sequencing_summary TEXT       Path to sequencing summary file. If
                                  provided, timestamps will be taken from this
                                  file.
  --browse                        Browse Historic Data.
  -e, --exclude [sturgeon|forest|nanodx|cnv|fusion|coverage|mgmt]
                                  Exclude analysis types with one or more of
                                  these options.
  --help                          Show this message and exit.
```

To run the tool, you will need to provide a watchfolder and an output folder. 

The watchfolder is where minKNOW will be writing the aligned BAM files for the run you wish to analyse.

The output folder is where results will be written too. This folder must exist and should be empty.

Optional flags include:

- --port: The port to run the GUI on.
- --threads: The number of threads to use for analysis. This will be the number of threads used by each tool. You should set this with care. On a system with a limited number of cores you should set this to 1. On systems with more CPU cores available you can use more cores. #ToDo: Sensibly manage cores between processes.
- --simtime: This introduces a delay in the adding of bam files to the pipeline and can be used if pointing cnsmeth at historic data.
- --showerrors: This will display all errors in R. This is useful for debugging. it should be explicitly set - i.e --showerros True
- --sequencing_summary: This is the path to a sequencing summary file. If provided, timestamps will be taken from this file and data will be loaded based on this file. 
- --browse: This will allow you to browse historic data. This feature is currently incomplete. #ToDo: Complete this feature!
- -e, --exclude: This will allow you to exclude certain analysis types. For example, if you do not wish to run any specific classifier just exclude it. the -e flag can be used multiple times.


A typical command line would look like this:

```console
cnsmeth --threads 4  /path/to/watchfolder /path/to/output
```

Upon launch, the GUI will open in your default browser. It should look something like this:

![img.png](images/img.png)

The interface will automatically update as new files are generated by MinKNOW.

The top panel provides information summarising general features of the run:

![img_1.png](images/img_1.png)

The next panel provides a summary of classification results. This will update as the run progresses and will be summarised in the top panel.

![img_2.png](images/img_2.png)

The next panel shows information on copy number changes - again this will update during the run, but is also interactive. Individual chromosomes can be inspected in more detail and specific genes highlighted.

![img_3.png](images/img_3.png)

![img_4.png](images/img_4.png)


The coverage panel shows per chromosome coverage and coverage for the targets and off target regions during sequencing. You can also visualise the change in coverage over time as well as the coverage for specific targets.

![img_5.png](images/img_5.png)

The next panel shows the methylation status across the MGMT promoter region. This plot will take considerable time to generate sufficient coverage to be meaningful.

![img_6.png](images/img_6.png)

The final panel shows the fusion gene status. This will update as the run progresses and will be summarised in the top panel. High confident fusions occur between two genes in the target panel whereas low confidence fusions occur between a gene in the target panel and a gene elsewhere in the genome. Candidate fusions are identified on the basis of scanning data for reads with supplementary alignments.

![img_7.png](images/img_7.png)





## About

This tool uses a range of third party tools and applications including:

- [Sturgeon] https://github.com/marcpaga/sturgeon
- [Radid-CNS2] https://link.springer.com/article/10.1007/s00401-022-02415-6
- [Readfish] https://github.com/LooseLab/readfish
- [cnv_from_bam] https://github.com/adoni5/cnv_from_bam
- [methylartist] https://github.com/adamewing/methylartist

We are grateful to the authors of these tools for their work.

We also thank a lot of people who have contributed to these tools including: Graeme Fox, Simon Deacon, Rory Munro, Satrio Wibowo, Thomas Murray, Inswasti Cahyani, Nadine Holmes, Simon Paine, Stuart Smith and many others from outside Nottingham.

We are particularly grateful to Areeba Patel, Felix Sahm and colleagues for their work on Rapid-CNS2.

This list is non-exhaustive and the software is under active development.

Documentation is currently unavailable.

This software is provided "as is", and is for research use only.


## License

`CNSmeth` is distributed under the terms of the [MIT](https://spdx.org/licenses/MIT.html) license.
